ouroboros-consensus/test-consensus/Test/Consensus/Protocol/PBFT.hs

```diff
@@ -24,220 +25,312 @@ import           Ouroboros.Network.Block (SlotNo (..))
+    , testChainInputsC       :: Inputs PBftMockCrypto
 
-      -- | The slots that were dropped (rolled back to create 'testChainState')
-    , testChainDroppedInputs  :: Inputs PBftMockCrypto
+      -- | @csABb@
+    , testChainState         :: PBftChainState PBftMockCrypto
 
-      -- | Possible next inputs that can be added to 'testChainState' and to
-      -- 'testChainOldState'
-      --
-      -- INVARIANT: the length of this list will be @n + k@.
-    , testChainNextInputs     :: Inputs PBftMockCrypto
+      -- | The @csABb@ we would have deserialised before #1307: this state does
+      -- not include the @n@ pre-anchor signatures.
+    , testChainOldState      :: PBftChainState PBftMockCrypto
 
-      -- | Possible next rollback from 'testChainState'
-    , testChainRollback       :: WithOrigin SlotNo
+      -- | The slot of some input within segment A
+    , testChainRewind        :: WithOrigin SlotNo
 
-      -- | The inputs that 'testChainRollback' would roll back
-    , testChainRollbackInputs :: Inputs PBftMockCrypto
+      -- | The inputs in segment A occupying slots @> 'testChainRewind'@
+    , testChainRewoundInputs :: Inputs PBftMockCrypto
     }
   deriving (Show)
 
--- | Generate chain state
---
--- We don't want to make too many assumptions here, so instead we model what
--- happens on an actual chain: we have a bunch of slots, starting from genesis,
--- signed by various keys, some slots are empty. We then compute the maximum
--- rollback point (@k@ from the tip), and finally drop (d @<= k@) slots,
--- and take the final @n + k - d@ slots.
---
--- Returns both the chain before rollback as well as the chain state.
---
--- NOTE: PBftChainState assumes @k >= 1@.
 instance Arbitrary TestChainState where
-  arbitrary = do
-      k       <- choose (1, 4)
-      n       <- choose (1, 10)
-      numKeys <- choose (1, 4)
-
-      -- Pick number of signed slots
-      --
-      -- We don't try to be too clever here; we need to test the various edge
-      -- cases near genesis more carefully, but also want to test where we are
-      -- well past genesis
-      numSigners <- oneof [
-          choose (0, k)
-        , choose (0, n)
-        , choose (0, k + n)
-        , choose (0, 5 * (k + n))
+  arbitrary = genTestChainState
+
+genTestChainState :: Gen TestChainState
+genTestChainState = do
+    k       <- choose (1, 4)    -- security parameter
+    n       <- choose (1, 10)   -- PBFT window size
+    numKeys <- choose (1, 4)    -- number of core nodes
+
+    let paramK = SecurityParam k
+        paramN = CS.WindowSize n
+        genKey = genMockKey numKeys
+
+    -- Pick number of signed slots
+    --
+    -- We don't try to be too clever here; we need to test the various edge
+    -- cases near genesis more carefully, but also want to test where we are
+    -- well past genesis
+    numSignersAB <- oneof [
+        choose (0, k + 1)
+      , choose (0, n + 1)
+      , choose (0, k + n + 1)
```

Comment:
Why the `+ 1`s? Is it to cover the other side of the interesting bound as well?

Reply:
Yep; Edsko's comment said "edge cases", which made me think of off-by-ones.